core/src/main/java/org/elasticsearch/cluster/health/ClusterStateHealth.java

```diff
@@ -0,0 +1,240 @@
+
+        for (String index : concreteIndices) {
+            IndexRoutingTable indexRoutingTable = clusterState.routingTable().index(index);
+            IndexMetaData indexMetaData = clusterState.metaData().index(index);
+            if (indexRoutingTable == null) {
+                continue;
+            }
+
+            ClusterIndexHealth indexHealth = new ClusterIndexHealth(indexMetaData, indexRoutingTable);
+
+            indices.put(indexHealth.getIndex(), indexHealth);
+        }
+
+        status = ClusterHealthStatus.GREEN;
+
+        for (ClusterIndexHealth indexHealth : indices.values()) {
+            activePrimaryShards += indexHealth.getActivePrimaryShards();
+            activeShards += indexHealth.getActiveShards();
+            relocatingShards += indexHealth.getRelocatingShards();
+            initializingShards += indexHealth.getInitializingShards();
+            unassignedShards += indexHealth.getUnassignedShards();
+            if (indexHealth.getStatus() == ClusterHealthStatus.RED) {
+                status = ClusterHealthStatus.RED;
+            } else if (indexHealth.getStatus() == ClusterHealthStatus.YELLOW && status != ClusterHealthStatus.RED) {
+                status = ClusterHealthStatus.YELLOW;
+            }
+        }
+
+        if (!validationFailures.isEmpty()) {
+            status = ClusterHealthStatus.RED;
+        } else if (clusterState.blocks().hasGlobalBlock(RestStatus.SERVICE_UNAVAILABLE)) {
+            status = ClusterHealthStatus.RED;
+        }
+
+        // shortcut on green
+        if (status.equals(ClusterHealthStatus.GREEN)) {
+            this.activeShardsPercent = 100;
+        } else {
+            List<ShardRouting> shardRoutings = clusterState.getRoutingTable().allShards();
+            int activeShardCount = 0;
+            int totalShardCount = 0;
+            for (ShardRouting shardRouting : shardRoutings) {
+                if (shardRouting.active()) activeShardCount++;
+                totalShardCount++;
+            }
+            this.activeShardsPercent = (((double) activeShardCount) / totalShardCount) * 100;
+        }
+    }
+
+    public List<String> getValidationFailures() {
+        return Collections.unmodifiableList(validationFailures);
+    }
+
+    public int getActiveShards() {
+        return activeShards;
+    }
+
+    public int getRelocatingShards() {
+        return relocatingShards;
+    }
+
+    public int getActivePrimaryShards() {
+        return activePrimaryShards;
+    }
+
+    public int getInitializingShards() {
+        return initializingShards;
+    }
+
+    public int getUnassignedShards() {
+        return unassignedShards;
+    }
+
+    public int getNumberOfNodes() {
+        return this.numberOfNodes;
+    }
+
+    public int getNumberOfDataNodes() {
+        return this.numberOfDataNodes;
+    }
+
+    public ClusterHealthStatus getStatus() {
+        return status;
+    }
+
+    public void setStatus(ClusterHealthStatus status) {
```

Comment:
can we call it overrideStatus and add java docs? Also if you feel strongly about it, you can only do this in the 2.x back port and do something else on master.